Configurable authentication converter for resource-servers with token introspection

ch4mpy opened the pull request:
This could close https://github.com/spring-projects/spring-security/issues/11661

It adds configurable authentication converter for resource-servers with token introspection (something very similar to what `JwtAuthenticationConverter` does for resource-servers with JWT decoder).

The new `(Reactive)OpaqueTokenAuthenticationConverter` is given responsibility for converting successful token introspection result into an `Authentication` instance (which is currently done by private methods of `OpaqueTokenAuthenticationProvider` and `OpaqueTokenReactiveAuthenticationManager`).

`(Reactive)BearerTokenAuthenticationConverter`, the default `(Reactive)OpaqueTokenAuthenticationConverter`, behave the same as current private `convert(OAuth2AuthenticatedPrincipal principal, String token)` methods: map authorities from `scope` attribute and return a `BearerTokenAuthentication`.

New unit test classes are created for default authentication converters and tests related to authorities mapping are moved there.

Existing tests  for `OpaqueTokenAuthenticationProvider` and `OpaqueTokenReactiveAuthenticationManager` are modified to simply assert that introspector and authenticationConverter are called and that resulting Authentication is propagated.

P.S.
I set 6.0 as target but can of course modify this to 5.8 or whatever depending on how you schedule the feature (additional XSDs could need an update depending on the branch this is merged to).

ch4mpy commented:
@sjohnr and @marcusdacoregio I believe this PR to be ready for review.

sjohnr commented:
Thanks @ch4mpy! As we're working through some priority tasks for 5.8 and 6.0, it may be a slight delay before I'm able to take a look at this.

sjohnr commented:
Thanks for taking the time! We use the duplicate label for book-keeping so only one issue (either the issue or the PR) goes into the release notes, and so whoever did the work gets credit. Since you opened the issue AND the PR, I just marked this PR as the duplicate.

ch4mpy commented:
@sjohnr I added a separate commit for 5.8 XSD and doc, just in case you have some time to review any time soon (the commit is easy to remove otherwise)

sjohnr commented:
Thanks @ch4mpy. Apologies, I haven't been able to review it thoroughly yet, as I'm still heads down. However, glancing at it, I wanted to mention one thing to think about. You may want to consider the `Converter` interface from spring-core. It's used fairly extensively throughout the framework already, and we'd probably want to prefer that over introducing new interfaces. Switching that over will get you a bit ahead on review feedback.

I'll let you know when I have a bit more time to review.

ch4mpy commented:
Actually, I have considered using the `Converter` interface but had two reserves :
- current convertion methods have two parameters (token string and authenticated principal) => I'd have to create a new class, let's say `IntrospectionResult` to wrap the two parameters into a single one.
- due to generics erasure, bean factory happens to lost and I frequently have to curry some Converter interfaces (like `ABConverter extends Converter<A, B>`) for beans to be correctly wired... 

Let me know if you prefer the solution with the wrapping class and a converter, I'll update the PR.

sjohnr commented:
> Please confirm that IntrospectionResult is fine for naming the converter source type.

I'll spend some time thinking about that.

sjohnr reviewed (changes requested):
Thanks for the PR @ch4mpy! I'm excited about this one. As I mentioned, I'm still a little behind on tasks, but I wanted to give this a bit of attention. See below comments for some suggestions to get you moving forward.

Also, I mention this below but putting it at the top so you can focus on it first:

> At the moment, I'm re-thinking my position on using a Converter here. Would you mind reverting the changes to switch to this interface, so we have an OpaqueTokenAuthenticationConverter again? (Sorry for being indecisive!)

ch4mpy commented on config/src/main/resources/org/springframework/security/config/spring-security-6.0.xsd:
```diff
@@ -1277,7 +1277,7 @@
          <xs:annotation>
             <xs:documentation>Optional attribute that specifies that the SecurityContext should require explicit saving
-                rather than being synchronized from the SecurityContextHolder. Defaults to "false".
+                rather than being synchronized from the SecurityContextHolder. Defaults to "true".
```
It came with a rebase...

sjohnr replied:
Gotcha. You may need to revert this change on your branch and squash the commit, so your diff does not include this change. Then I think you should be able to rebase again.

ch4mpy replied:
I don't know exactly how, but I think I picked a version from main just before a regression: because of gh-11762, default value became "true". This can be seen in `spring-security-6.0.rnc` and more importantly in `HttpConfigurationBuilder.java` line 326: `return !StringUtils.hasText(explicitSaveAttr) || Boolean.parseBoolean(explicitSaveAttr);`

Do you really prefer I remove this change from this PR and create an issue just for this documentation line?

ch4mpy replied:
I will revert to false, but rnc and xsd won't be in synch. I already squashed and commited.

ch4mpy replied:
Created a separate pull request: https://github.com/spring-projects/spring-security/pull/11773

sjohnr replied:
Ok, sorry I didn't realize they were out of sync.

ch4mpy commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java:
```diff
@@ -68,13 +70,18 @@ public final class OpaqueTokenAuthenticationProvider implements AuthenticationPr
 
 	private final OpaqueTokenIntrospector introspector;
 
+	private final Converter<IntrospectionResult, ? extends Authentication> authenticationConverter;
```
@sjohnr sure, I can. Which option do you prefer:
1. `public interface OpaqueTokenAuthenticationConverter extends Converter<IntrospectionResult, ? extends Authentication>`
2. `public interface OpaqueTokenAuthenticationConverter<T extends Authentication> { T convert(String introspectedToken, OAuth2AuthenticatedPrincipal authenticatedPrincipal); }

I prefer option 2

ch4mpy commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/BearerTokenAuthenticationConverter.java:
```diff
@@ -0,0 +1,63 @@
+ * @author Jerome Wacongne &lt;ch4mp@c4-soft.com&gt;
+ * @since 5.8
+ */
+public class BearerTokenAuthenticationConverter implements Converter<IntrospectionResult, BearerTokenAuthentication> {
```
What  about `public class BearerTokenAuthenticationConverter implements OpaqueTokenAuthenticationConverter<BearerTokenAuthentication>` ?

ch4mpy commented:
@sjohnr I:
- rebased (twice, was lucky enough to pick an unstable main the first time)
- reverted to `OpaqueTokenAuthenticationConverter` interface and made it a `@FunctionalInterface`
- found a way to share code between servlet and reactive confs for default `OpaqueTokenAuthenticationConverter`
- added a test to assert XML conf for optional `OpaqueTokenAuthenticationConverter` is actually used
- removed breaking change by adding constructors with `OpaqueTokenAuthenticationConverter` argument (and keeping constructors without such an argument)

ch4mpy commented on docs/modules/ROOT/pages/servlet/oauth2/resource-server/opaque-token.adoc:
```diff
@@ -297,11 +297,13 @@ fun introspector(): OpaqueTokenIntrospector {
-Or, if you're not using Spring Boot at all, then both of these components - the filter chain and a <<oauth2resourceserver-opaque-architecture-introspector,`OpaqueTokenIntrospector`>> can be specified in XML.
+If the application doesn't expose an `OpaqueTokenAuthenticationConverter` bean, then Spring Boot will expose a `BearerTokenAuthenticationConverter`.
+
+Or, if you're not using Spring Boot at all, then all of these components - the filter chain, an <<oauth2resourceserver-opaque-architecture-introspector,`OpaqueTokenIntrospector`>> and an `OpaqueTokenAuthenticationConverter` can be specified in XML.
```
I tried to rephrase, but I'm not native English speaker. Please provide me with what you want if it still not what you expect.

sjohnr replied:
Ok. Will do!

ch4mpy commented:
Rebased on main after https://github.com/spring-projects/spring-security/pull/11773 was merged and squashed again.

ch4mpy commented:
@sjohnr, do you expect more from me?

sjohnr commented on oauth2/oauth2-resource-server/src/test/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProviderTests.java:
```diff
@@ -50,58 +45,37 @@ public class OpaqueTokenAuthenticationProviderTests {
-	}
 
-	@Test
-	public void authenticateWhenMissingScopeAttributeThenNoAuthorities() {
```
I'd prefer duplication in tests to refactoring existing tests. Can this test be restored to original form?

sjohnr reviewed (changes requested):
This is definitely developing nicely @ch4mpy! Let's leave the existing interface name for now.

I have more suggestions to some specific areas of the PR, see below. Once we work through this round, any additional (small) changes could be applied as polish by the team when we're ready to merge. Please note that we will pause at that point and wait on this PR until after November (after the 5.8/6.0 release).

ch4mpy commented:
> we will pause at that point and wait on this PR until after November (after the 5.8/6.0 release).

So I should remove modifications to 5.8 xsd and rnc too, maybe? And also modify `@Since` comments. Do you know what the target version will be?

ch4mpy commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java:
```diff
@@ -116,11 +135,21 @@ public boolean supports(Class<?> authentication) {
-		Instant exp = principal.getAttribute(OAuth2TokenIntrospectionClaimNames.EXP);
-		OAuth2AccessToken accessToken = new OAuth2AccessToken(OAuth2AccessToken.TokenType.BEARER, token, iat, exp);
-		return new BearerTokenAuthentication(principal, accessToken, principal.getAuthorities());
+	public static BearerTokenAuthentication convert(String introspectedToken,
```
Package visibility is a minimum to share code between servlet and reactive versions

ch4mpy commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java:
```diff
@@ -116,11 +135,21 @@ public boolean supports(Class<?> authentication) {
+				authenticatedPrincipal.getAuthorities());
+	}
+
+	public OpaqueTokenAuthenticationProvider setAuthenticationConverter(
```
No, it doesn't need, but does it remove usability?

Will convert to void return.

sjohnr replied:
It doesn't impair usability, the suggestion was more for consistency. :smile:

ch4mpy commented:
@sjohnr I think I fixed everything as you requested, but `static BearerTokenAuthentication convert(String introspectedToken, OAuth2AuthenticatedPrincipal authenticatedPrincipal)` for which I set visibility to package instead of private to keep code de-duplication

sjohnr commented:
> So I should remove modifications to 5.8 xsd and rnc too, maybe? And also modify @Since comments. Do you know what the target version will be?

We can wait on that, but thanks for asking! I'll try and keep you updated if anything changes.

sjohnr commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/authentication/OpaqueTokenAuthenticationProvider.java:
```diff
@@ -92,8 +100,7 @@ public Authentication authenticate(Authentication authentication) throws Authent
 		BearerTokenAuthenticationToken bearer = (BearerTokenAuthenticationToken) authentication;
 		OAuth2AuthenticatedPrincipal principal = getOAuth2AuthenticatedPrincipal(bearer);
-		AbstractAuthenticationToken result = convert(principal, bearer.getToken());
-		result.setDetails(bearer.getDetails());
```
One thing I just noticed, we should ensure the details are still being set if applicable. I believe it would be acceptable to introduce an check for `AbstractAuthenticationToken` and cast/set the details if true. Similar to:

https://github.com/spring-projects/spring-security/blob/d996c2a2c6d14c4e578cb56019737922060fd7da/saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/servlet/filter/Saml2WebSsoAuthenticationFilter.java#L138-L143

I'm currently preferring that instead of returning `AbstractAuthenticationToken` from our converter.

**Note:** We don't need to introduce a `AuthenticationDetailsSource` at this time.

ch4mpy replied:
I'm not found of the idea to alter the Authentication instance after the authenticationConverter returned it. Maybe limit that to just `BearerAuthenticationToken` instances instead of all `AbstractAuthenticationToken`?

ch4mpy replied:
I think I found a way to do it from inside default converter. Please let me know what you think of my latest commit.

sjohnr replied:
Thanks @ch4mpy, it's an interesting idea! However, I'm not so sure in this case. As far as I can tell, this solution only works when the default converter is used, and introduces more code to do so. Doing a quick search for `setDetails` in the code base, it's fairly common for a filter or provider to set the details afterwards. I think adding a check and using `AbstractAuthenticationToken` provides good flexibility and consistency with the rest of the framework.

ch4mpy replied:
> this solution only works when the default converter is used

This is exactly my intention. Concider this sample (it compiles)
```java
	public static class KeycloakClaims extends HashMap<String, Object> implements Serializable, StandardClaimAccessor {
		@Override
		public Map<String, Object> getClaims() {
			return this;
		}
		
		public List<String> getRoles() {
			final var realmAccess = (Map<String, Object>) getOrDefault("realm_access", Map.of());
			return (List<String>) realmAccess.getOrDefault("roles", List.of());
		}
	}
	
	public static class RealmAuthentication extends AbstractAuthenticationToken {

		public RealmAuthentication(KeycloakClaims claims) {
			super(claims.getRoles().stream().map(SimpleGrantedAuthority::new).toList());
			this.setDetails(claims);
		}

		@Override
		public void setDetails(Object details) {
			assert(KeycloakClaims.class.isAssignableFrom(details.getClass()));
			super.setDetails(details);
		}

		@Override
		public KeycloakClaims getDetails() {
			return (KeycloakClaims)super.getDetails();
		}

		@Override
		public KeycloakClaims getCredentials() {
			return this.getDetails();
		}

		@Override
		public KeycloakClaims getPrincipal() {
			return this.getDetails();
		}
	}
```
Clearly, trying to set details with a random `Object` after the authenticationConverter returned such an `AbstractAuthenticationToken` implementation will be a problem.

The aim of this request is to give total control over `Authentication` instance to the framework user after successful introspection.

I understand the need to keep exact same behaviour for default authenticationConverter, but I think it would be very tricky (and frustrating) for one providing explicitly an authenticationConverter to have the authentication instance it returned be arbitrarly altered (with value out of his control).

## PS
As far I can read from existing code, `details` are set for servlets only. Is it intended or just a bug introduced when duplicating code?

sjohnr replied:
Those are good points, @ch4mpy.

> The aim of this pull request is to give total control over `Authentication` instance to the framework user after successful introspection.

It may be that a converter is not the best solution in that case? Introducing a converter adds a configuration hook that gives flexibility to the user. If total control is the goal, there are other ways to accomplish that. For example, we could build a delegating `AuthenticationProvider`:

```java
public class MyOpaqueTokenAuthenticationProvider implements AuthenticationProvider {

	private final AuthenticationProvider delegate;

	public MyOpaqueTokenAuthenticationProvider(OpaqueTokenIntrospector introspector) {
		this.delegate = new OpaqueTokenAuthenticationProvider(introspector);
	}

	@Override
	public boolean supports(Class<?> authentication) {
		return this.delegate.supports(authentication);
	}

	@Override
	public Authentication authenticate(Authentication authentication) throws AuthenticationException {
		Authentication result = this.delegate.authenticate(authentication);
		// ...
	}

}
```

This gives full control at the expense of constructing a new authentication based on the returned one. Of course, you can always choose to fully implement the `AuthenticationProvider` yourself if needed.

Having said that, it sounds like my suggestion conflicts with your intended use case for introducing the converter? If that's the case, perhaps your authentication instance shouldn't extend from `AbstractAuthenticationToken`? Or it could simply store the claims in another field, which seems to make sense here. The claims in your example seem more specific than `Object details` from `AbstractAuthenticationToken`.

> As far I can read from existing code, `details` are set for servlets only. Is it intended or just a bug introduced when duplicating code in `OpaqueTokenReactiveAuthenticationManager`?

I'd have to dig into it to say for 100% certain, but it could be that the use of `details` addresses a servlet-specific need.

ch4mpy replied:
> The claims in your example seem more specific than `Object details` from `AbstractAuthenticationToken`.

Hopefully, yes, it is. To be honest, I like typed programming languages and wish there were no `Object` return type in the framework (`Authentication` and `AbstractAuthenticationToken` could use generics or something), but this is not the subject here.

> perhaps your authentication instance shouldn't extend from AbstractAuthenticationToken?

Implementing just `Authentication` would keep such an implementation away of all the part of the framework which expect `AbstractAuthenticationToken`?

> could simply store the claims in another field

Yes, maybe. I just like the idea of specializations (`B extends A`) to be more specific, specially when `A` is way too generic (like having an `Object` return type)

If you really want to keep the side effect on details for all `AbstractAuthenticationToken`, inside servlets only, then maybe should it be carefully documented. The `Authentication` implementations I write can live with it and my security expressions will just omit `details` property (if I don't know what is in there, I will certainely not use it).

Are you sure you don't want the test to apply the `details` side effect to be applied on `OpaqueTokenAuthentication` instances only instead of all `AbstractAuthenticationToken`?

sjohnr replied:
Thanks for explaining your point of view, @ch4mpy. I think if we were designing a new component entirely (new authentication provider or filter) we would have quite a bit more license to use these ideas to define it ourselves.

> Are you sure you don't want the `details` side effect to be applied to `BearerTokenAuthentication` instances only (instead of all `AbstractAuthenticationToken`)?

This is not what the `AuthenticationProvider` does today, so technically this would be a breaking change to the default behavior, which is mainly what I'm proposing we try and avoid.

It's worth noting that once this authentication is returned to the framework, the `ProviderManager` does exactly what I'm proposing but with a `null` check:

https://github.com/spring-projects/spring-security/blob/929b3345886c1810b32cd38257bf19c629008b92/core/src/main/java/org/springframework/security/authentication/ProviderManager.java#L258-L263

If you want to add the same `null` check and not overwrite what was set in the converter, I think that would be a fair compromise and wouldn't be a breaking change. What do you think?

ch4mpy replied:
Yes, I prefer to add a null check to set details only if was "forgotten".

Should this be applied to reactive apps too?

sjohnr commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/ReactiveOpaqueTokenAuthenticationConverter.java:
```diff
@@ -0,0 +1,35 @@
+@FunctionalInterface
+public interface ReactiveOpaqueTokenAuthenticationConverter {
+
+	Mono<Authentication> convert(String introspectedToken, OAuth2AuthenticatedPrincipal authenticatedPrincipal);
```
I think we should also add javadoc here.

sjohnr commented on oauth2/oauth2-resource-server/src/main/java/org/springframework/security/oauth2/server/resource/introspection/OpaqueTokenAuthenticationConverter.java:
```diff
@@ -0,0 +1,33 @@
+@FunctionalInterface
+public interface OpaqueTokenAuthenticationConverter {
+
+	Authentication convert(String introspectedToken, OAuth2AuthenticatedPrincipal authenticatedPrincipal);
```
I think we should also add javadoc here.

sjohnr reviewed (changes requested):
It's looking really good @ch4mpy! I did find a few small things below. Sorry for missing them earlier!

sjohnr commented:
Thanks for the updates @ch4mpy. After discussing with the team, I think we're good to merge PRs from the community into 5.8. As I mentioned, we've switched to merge-forward as a strategy, so we ultimately want to target 5.8 for this PR. If you happen to have a bit more time, you could squash commits and move the commit to the `5.8.x` branch to see if it moves without conflicts (I believe you can edit the base of this PR without opening a new one). If you don't have time, don't worry about it and I can try and find time in the next week or two to get it moved over.

ch4mpy commented:
@sjohnr I squashed the commits.

Unforunately, I forked https://github.com/spring-projects/spring-security with solely the `main` branch (forgot to uncheck the option when creating the fork, my bad).

This makes things for me a bit more complicated than just cherry-pick my squashed commit on 5.8.x branch :-/ 

If you don't mind, I'd appreciate someone with all branches do it for me. I'll then trash my fork and get a fresh one with all branches.

ch4mpy commented:
@sjohnr squashed and rebased on `5.8.x`.

Of course, 6.0 xsd and rnc modifications where lost in the process.
